SalesforceCommerceCloud/pwa-kit
# Unified configurations Approach #2 (reduced scope)

This is another approach for PR https://github.com/SalesforceCommerceCloud/pwa-kit/pull/185.

## Problem:

Developers don't understand the configuration options, and there is no schema that helps developer understand them. There needs to have validation to help developers catch configuration errors.

## Solution:

Created a config file `pwa-kit.config.json` at the root of the PWA project. This file contains all PWA configurations.

This PR adds a new webpack plugin `PwaKitConfigPlugin` to the SDK. The plugin is responsible for taking the configuration file and validate against the schema.

The webpack plugin takes 2 optional arguments:`path` and `customSchemas`, the `path` is used when the file is placed in a different location (default location is /pwa/pwa-kit.config.json) and `customSchemas` will be used to extend the configuration schema. The schema is added in the SDK and will be shipped as part of the package.

Finally we extend the webpack configuration in the user land in webpack.config.js.

We installed 2 new dependencies in this change: `ajv` and `ajv-merge-patch`, they are used to validate json objects.

## Test:
- clone the repo if you haven't done so, run `npm ci` at root
- `cd packages/pwa` and then `npm start`
- the app should start as normal since the default config file has correct format.
- change `pwa-kit.config.json`, update `app` to have value of 1 (number instead of an object)
- verify you get console warning like this:
<img width="405" alt="Screen Shot 2021-12-08 at 2 44 44 PM" src="https://user-images.githubusercontent.com/10948652/145303305-526e1b5d-e8a4-4878-81da-d8303e3ca2f7.png">
<img width="341" alt="Screen Shot 2021-12-08 at 2 23 37 PM" src="https://user-images.githubusercontent.com/10948652/145303356-2d8528de-1466-4ba9-956e-e66921d9b147.png">

- try to make the validation fail by changing the configuration values, e.g. remove required fields. you should see console errors.
- fix the configuration value, and the app server should restart and recover.
